### PR TITLE
Update SceneBuilder to check for required PointCloud fields.

### PR DESCRIPTION
**Pull request by defunctzombie**
**User-Facing Changes**
Rather than crashing the app when fields are missing from the PointCloud message, the user is shown a topic error next to the malformed topic.

image

Any other rendering errors result in a panel crash.

**Description**
PointCloud rendering requires field descriptions for x, y, and z fields. This change updates SceneBuilder to check for these fields prior to adding the point cloud to the collector. This allows the UI to display a topic error for the malformed topic better informing the user to errors with their data.

Fixes: #2759

### CODE REVIEW

File: packages/studio-base/src/panels/ThreeDimensionalViz/SceneBuilder/index.ts [hunk 1]
```diff
@@ -974,8 +974,34 @@ export default class SceneBuilder implements MarkerProvider {
         return add.triangleList(marker);
       case 101:
         return add.grid(marker);
-      case 102:
+      case 102: {
+        // PointCloud decoding requires x, y, and z fields and will fail if all are not present.
```
Comment (defunctzombie) [inline packages/studio-base/src/panels/ThreeDimensionalViz/SceneBuilder/index.ts:978]: Is this an acceptable price to pay? Probably no worse than all the other logic that happens for markers? I could also do this for the first message on each topic and assume follow-up messages are ok?

File: packages/studio-base/src/panels/ThreeDimensionalViz/SceneBuilder/index.ts [hunk 2]
```diff
@@ -974,8 +974,34 @@ export default class SceneBuilder implements MarkerProvider {
...
+        // We check for the fields here so we can present the user with a topic error prior to decoding.
+        const fieldNames: { [key: string]: boolean } = {};
```
Comment (jtbandes) [inline packages/studio-base/src/panels/ThreeDimensionalViz/SceneBuilder/index.ts:980]: `Set`?

File: packages/studio-base/src/panels/ThreeDimensionalViz/SceneBuilder/index.ts [hunk 3]
```diff
@@ -974,8 +974,34 @@ export default class SceneBuilder implements MarkerProvider {
...
+        if (fieldNames.y == undefined) {
+          missingFields += " y";
+        }
+        if (fieldNames.z == undefined) {
+          missingFields += " z";
+        }
+        if (missingFields.length > 0) {
+          this._setTopicError(
+            topic.name,
+            `Missing required fields entry for field names: ${missingFields}`,
```
Comment (jtbandes) [inline packages/studio-base/src/panels/ThreeDimensionalViz/SceneBuilder/index.ts:998]:
```suggestion
            `Point cloud is missing required fields: ${missingFields}`,
```

File: packages/studio-base/src/panels/ThreeDimensionalViz/commands/PointClouds/index.tsx [hunk 4]
```diff
@@ -444,30 +443,14 @@ type Props = CommonCommandProps & {
...
-    : (children as PointCloudMarker[])
-        .map((m) => {
-          try {
-            return decodeMarker(m);
-          } catch (err) {
-            sendNotification("Point cloud decoding failed", err, "user", "error");
-            return undefined;
-          }
-        })
-        .filter((m) => m != undefined);
```
Comment (jtbandes) [inline packages/studio-base/src/panels/ThreeDimensionalViz/commands/PointClouds/index.tsx:470]: So by removing this, we now crash the whole panel when a point cloud can't be rendered?

File: packages/studio-base/src/util/sendNotification.ts [hunk 5]
```diff
@@ -48,7 +48,7 @@ const defaultNotificationHandler: NotificationHandler = (
   type: NotificationType,
   severity: NotificationSeverity,
 ): void => {
-  if (!inWebWorker()) {
+  if (inWebWorker()) {
```
Comment (jtbandes) [inline packages/studio-base/src/util/sendNotification.ts:51]: :philosorapter:

**Review by jtbandes — APPROVED** [review]

File: packages/studio-base/src/panels/ThreeDimensionalViz/commands/PointClouds/index.tsx (hunk 4 above)
Comment (defunctzombie) [reply packages/studio-base/src/panels/ThreeDimensionalViz/commands/PointClouds/index.tsx:470]: Yes - I'm open to other ideas but IMO send notification is not the correct UX strategy.

File: packages/studio-base/src/panels/ThreeDimensionalViz/SceneBuilder/index.ts (hunk 2 above)
Comment (defunctzombie) [reply packages/studio-base/src/panels/ThreeDimensionalViz/SceneBuilder/index.ts:980]: Significantly slower /shrug

File: packages/studio-base/src/util/sendNotification.ts (hunk 5 above)
Comment (defunctzombie) [reply packages/studio-base/src/util/sendNotification.ts:51]: Yea - I did not understand this. Its like this in [webviz](https://github.com/cruise-automation/webviz/blob/master/packages/webviz-core/src/util/sendNotification.js#L38) as well. What am I missing?

File: packages/studio-base/src/panels/ThreeDimensionalViz/commands/PointClouds/index.tsx (hunk 4 above)
Comment (jhurliman) [reply packages/studio-base/src/panels/ThreeDimensionalViz/commands/PointClouds/index.tsx:470]: Should the PR description be updated to mention that all other pointcloud rendering errors have been converted into a panel crash?

File: packages/studio-base/src/util/sendNotification.ts (hunk 5 above)
Comment (jtbandes) [reply packages/studio-base/src/util/sendNotification.ts:51]: Probably nothing.

File: packages/studio-base/src/panels/ThreeDimensionalViz/commands/PointClouds/index.tsx (hunk 4 above)
Comment (defunctzombie) [reply packages/studio-base/src/panels/ThreeDimensionalViz/commands/PointClouds/index.tsx:470]: Updated the PR description to note this. I'm also up to hearing how else to handle this that doesn't involve sending a notification to the user on every attempt to render a frame.

**Comment by AndyYangjd** [comment]
The studio can't render the pointcloud. The error info is: non-identity pose is not yet supported.


Is there any config error?

**Pull request merged**